test/Dapr.E2E.Test/DaprTestApp.cs

```diff
@@ -31,28 +33,63 @@ public DaprTestApp(ITestOutputHelper output, string appId, bool useAppPort = fal
             this.testOutput = output;
         }
 
-        public (string, string) Start()
+        public string AppId => this.appId;
+
+        public (string httpEndpoint, string grpcEndpoint) Start()
         {
             var (appPort, httpPort, grpcPort, metricsPort) = GetFreePorts();
+
             var componentsPath = Combine(".", "..", "..", "..", "..", "..", "test", "Dapr.E2E.Test", "components");
-            var daprStartCommand = $" run --app-id {appId} --dapr-http-port {httpPort} --dapr-grpc-port {grpcPort} --metrics-port {metricsPort} --components-path {componentsPath}";
             var projectPath = Combine(".", "..", "..", "..", "..", "..", "test", "Dapr.E2E.Test.App", "Dapr.E2E.Test.App.csproj");
-            var daprDotnetCommand = $" -- dotnet run --project {projectPath} --framework {GetTargetFrameworkName()}";
+            var arguments = new List<string>()
+            {
+                // `dapr run` args
+                "run",
+                "--app-id", appId,
+                "--dapr-http-port", httpPort.ToString(CultureInfo.InvariantCulture),
+                "--dapr-grpc-port", grpcPort.ToString(CultureInfo.InvariantCulture),
+                "--metrics-port", metricsPort.ToString(CultureInfo.InvariantCulture),
+                "--components-path", componentsPath,
+                "--log-level", "debug",
+                
+            };
+
+            if (this.useAppPort)
+            {
+                arguments.AddRange(new[]{ "--app-port", appPort.ToString(CultureInfo.InvariantCulture), });
+            }
+
+            arguments.AddRange(new[]
+            {
+                // separator
+                "--",
+
+                // `dotnet run` args
+                "dotnet", "run",
+                "--project", projectPath,
+                "--framework", GetTargetFrameworkName(),
+
+                // Prevent building again, it slows things down as restore can be slow or unreliable in CI environments
+                "--no-build",
+                "--no-restore",
+            });
```

Comment:
I find this approach really preferrable for commandline args. it's easier to read and easy to add things like quoting if we need it.